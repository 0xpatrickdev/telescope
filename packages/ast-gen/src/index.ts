export * from './types';
export * from './amino';
export * from './client';
export * from './messages';
export * from './utils';