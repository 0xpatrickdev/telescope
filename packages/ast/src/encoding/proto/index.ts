export * from './decode';
export * from './encode';
export * from './enums';
export * from './types';
export * from './from-json';
export * from './from-sdk';
export * from './from-amino';
export * from './from-partial';
export * from './to-json';
export * from './to-sdk';
export * from './to-amino';
export * from './interface';