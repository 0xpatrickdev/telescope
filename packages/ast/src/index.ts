export * from './encoding';
export * from './clients';
export * from './registry';
export * from './services';
export * from './utils';
export * from './bundle';
export * from './types';