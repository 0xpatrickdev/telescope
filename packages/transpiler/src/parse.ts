import { ProtoStore, ProtoRoot, ProtoRef, getObjectName } from '@osmonauts/proto-parser';
import {
    createProtoType,
    createCreateProtoType,
    createProtoEnum,
    createProtoObjectWithMethods,
    ProtoType,
    ProtoParseContext,
    AminoParseContext,
    makeAminoTypeInterface,
    aminoConverter,
    getKeyTypeEntryName,
    // client
    createClient,
    GenericParseContext,
    // registry 
    createTypeRegistry,
    createRegistryLoader,
    // methods (needs refactor)
    toObjectWithPartialMethods,
    toObjectWithEncodedMethods,
    toObjectWithJsonMethods,
    toObjectWithToJSONMethods,
    toObjectWithFromJSONMethods
} from '@osmonauts/ast';
import { camel, snake } from 'case';

const getRoot = (ref: ProtoRef): ProtoRoot => {
    if (ref.traversed) return ref.traversed;
    return ref.proto;
};

export interface TelescopeParseContext {
    proto: ProtoParseContext;
    amino: AminoParseContext;
    store: ProtoStore;
    ref: ProtoRef;
    parsedImports: Record<string, any>;
    body: any[];
    mutations: any[];
    queries: any[];
    types: any[];
}

export const parse = (
    context: TelescopeParseContext,
): void => {
    const root = getRoot(context.ref);
    parseRecur({
        context,
        obj: root.root,
        scope: [],
        isNested: false
    });
};

export const getParsedObjectName = (
    ref: ProtoRef,
    obj: any,
    scope: string[],
) => {
    const root = getRoot(ref);
    const allButPackage = scope.splice(root.package.split('.').length);
    // pull off "this" name
    allButPackage.pop();
    return getObjectName(obj.name, [root.package, ...allButPackage]);
};

// TODO potentially move this back to ast or proto bc the ast lib references MapEntries...
const makeKeyTypeObj = (ref: ProtoRef, field: any, scope: string[]) => {
    const root = getRoot(ref);
    const scoped = scope.splice(root.package.split('.').length);
    const adhocObj: ProtoType = {
        comment: undefined,
        fields: {
            key: {
                id: 1,
                type: field.keyType,
                scope: [...scoped],
                parsedType: {
                    name: field.keyType,
                    type: field.keyType
                },
                comment: undefined,
                options: undefined
            },
            value: {
                id: 2,
                type: field.parsedType.name,
                scope: [...scoped],
                parsedType: {
                    name: field.type,
                    type: field.parsedType.type
                },
                comment: undefined,
                options: undefined
            }
        }
    };
    return adhocObj;
}

export const parseType = (
    context: TelescopeParseContext,
    obj: any,
    scope: string[],
    isNested: boolean = false
) => {

    obj.keyTypes.forEach(field => {
        const keyTypeObject = makeKeyTypeObj(context.ref, field, scope);
        const name = getParsedObjectName(context.ref, {
            name: getKeyTypeEntryName(obj.name, field.parsedType.name)
        }, scope);
        context.body.push(createProtoType(name, keyTypeObject));
        context.body.push(createCreateProtoType(name, keyTypeObject));
        context.body.push(createProtoObjectWithMethods(context.proto, name, keyTypeObject));
    })

    // parse nested names
    let name = obj.name;
    if (isNested) {
        name = getParsedObjectName(context.ref, obj, scope);
    }

    // context for types that can be exported
    if (!isNested) {
        context.types.push(
            {
                [name]: obj
            }
        );
    }

    context.body.push(createProtoType(name, obj));
    context.body.push(createCreateProtoType(name, obj));
    context.body.push(createProtoObjectWithMethods(context.proto, name, obj));

    // render nested LAST
    if (obj.nested) {
        Object.keys(obj.nested).forEach(key => {
            // isNested = true;
            parseRecur({
                context,
                obj: obj.nested[key],
                scope: [...scope, key],
                isNested: true
            });
        });
    }

};

export const parseEnum = (
    context: TelescopeParseContext,
    obj: any,
    scope: string[],
    isNested: boolean = false
) => {
    let name = obj.name;
    // parse nested names
    if (isNested) {
        name = getParsedObjectName(context.ref, obj, scope);
    }
    context.body.push(createProtoEnum(name, obj));
};

export const parseService = (
    context: TelescopeParseContext,
    obj: any,
    scope: string[],
    isNested: boolean = false
) => {

    const methodHash: Record<string, {
        requestType: string;
        responseType: string;
        comment?: string;
    }> = obj.methods;

    if (!['Msg', 'Query'].includes(obj.name)) {
        throw new Error('unsupported Service type. Contact Maintainers.');
    }

    const isMutation = obj.name === 'Msg';

    // NOTE
    // in future this can get hoisted up and we could theoretically aggregate multiple 
    // proto files in the same package for same-named services, e.g. two files with `service Msg`

    const lookups = Object.entries(methodHash)
        .map(([key, value]) => {
            const lookup = context.store.get(context.ref, value.requestType);
            if (!lookup) {
                console.warn(`cannot find ${value.requestType}`);
                throw new Error('undefined symbol for service.');
            }
            const lookupResponse = context.store.get(context.ref, value.responseType);
            if (!lookupResponse) {
                console.warn(`cannot find ${value.requestType}`);
                throw new Error('undefined symbol for service.');
            }
            (isMutation ? context.mutations : context.queries).push({
                methodName: key,
                package: context.ref.proto.package,
                message: lookup.importedName,
                messageImport: lookup.import ?? context.ref.filename,
                response: lookupResponse.importedName,
                responseImport: lookupResponse.import ?? context.ref.filename,
                comment: value.comment
            })
            return lookup;
        });


    if (!lookups.length) return;

    const methods = Object.entries(methodHash)
        .map(([key, value]) => {
            return {
                typeUrl: `/${context.ref.proto.package}.${value.requestType}`,
                TypeName: value.requestType,
                methodName: camel(key)
            }
        });

    // AMINO INTERFACES
    lookups.forEach(lookup => {
        const proto = lookup.obj;
        context.body.push(makeAminoTypeInterface({
            context: context.amino,
            proto,
            options: {
                aminoCasingFn: snake
            }
        }));
    });
    const protos = lookups.map(lookup => lookup.obj);

    // AMINO CONVERTER
    context.body.push(aminoConverter({
        name: 'AminoConverter',
        context: context.amino,
        root: context.ref.traversed,
        protos,
        options: {
            aminoCasingFn: snake
        }
    }))

    // add registry

    context.body.push(createTypeRegistry(methods));
    context.body.push(createRegistryLoader());

    // add methods
    context.body.push(
        toObjectWithPartialMethods(methods)
    );
    context.body.push(
        toObjectWithEncodedMethods(methods)
    );
    context.body.push(
        toObjectWithJsonMethods(methods)
    );
    context.body.push(
        toObjectWithToJSONMethods(methods)
    );
    context.body.push(
        toObjectWithFromJSONMethods(methods)
    );
    context.body.push(
        toObjectWithEncodedMethods(methods)
    );

    // add client

    // NEED BUNDLE
    // BUNDLE needs planning FILES/FILE structure

};

interface ParseRecur {
    context: TelescopeParseContext,
    obj: any;
    scope: string[];
    isNested: boolean;
}
export const parseRecur = ({
    context,
    obj,
    scope,
    isNested
}: ParseRecur) => {
    switch (obj.type) {
        case 'Type':
            return parseType(
                context, obj, scope, isNested
            );
        case 'Enum':
            return parseEnum(
                context, obj, scope, isNested
            );
        case 'Service':
            return parseService(
                context, obj, scope, isNested
            );
        case 'Field':
            console.log(obj);
            return;
        case 'Root':
        case 'Namespace':
            if (obj.nested) {
                return Object.keys(obj.nested).forEach(key => {
                    parseRecur({
                        context,
                        obj: obj.nested[key],
                        scope: [...scope, key],
                        isNested
                    });
                });
            } else {
                throw new Error('parseRecur() cannot find protobufjs Type')
            }
        default:
            console.log({ obj });
            throw new Error('parseRecur() cannot find protobufjs Type')
    }
};