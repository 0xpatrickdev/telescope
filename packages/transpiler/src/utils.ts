import { ProtoStore, ProtoRoot, ProtoRef, getObjectName } from '@osmonauts/proto-parser';
import { relative, dirname, extname } from 'path';

export const getRoot = (ref: ProtoRef): ProtoRoot => {
    if (ref.traversed) return ref.traversed;
    return ref.proto;
};

export const UTILS = {
    AminoHeight: '@osmonauts/helpers',
    AminoMsg: '@cosmjs/amino',
    AminoTypes: '@cosmjs/stargate',
    decodeBech32Pubkey: '@cosmjs/amino',
    defaultRegistryTypes: '@cosmjs/stargate',
    encodeBech32PubKey: '@cosmjs/amino',
    fromBase64: '@cosmjs/encoding',
    fromBech32: '@cosmjs/encoding',
    fromDuration: '@osmonauts/helpers',
    fromHex: '@cosmjs/encoding',
    fromJsonTimestamp: '@osmonauts/helpers',
    fromTimestamp: '@osmonauts/helpers',
    GeneratedType: '@cosmjs/proto-signing',
    isSet: '@osmonauts/helpers',
    Long: '@osmonauts/helpers', // exports Long and also calls the magic Long code
    // Long: {type: 'default', 'long'}
    OfflineSigner: '@cosmjs/proto-signing',
    omitDefault: '@osmonauts/helpers',
    Registry: '@cosmjs/proto-signing',
    SigningStargateClient: '@cosmjs/stargate',
    toBase64: '@cosmjs/encoding',
    toDuration: '@osmonauts/helpers',
    toTimestamp: '@osmonauts/helpers',
    bytesFromBase64: '@osmonauts/helpers',
    base64FromBytes: '@osmonauts/helpers',
    _m0: 'protobuf/minimal'
    // _m0: { type: 'namespace', path: 'protobuf/minimal' }
};


export const getRelativePath = (f1: string, f2: string) => {
    const rel = relative(dirname(f1), f2);
    let importPath = rel.replace(extname(rel), '');
    if (!/\//.test(importPath)) importPath = `./${importPath}`;
    return importPath;
}

export const insertUniq = (array: string[], el: string) => {
    if (!array.includes(el)) {
        array.push(el);
    }
}
