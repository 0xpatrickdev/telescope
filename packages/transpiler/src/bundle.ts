import { ProtoStore } from '@osmonauts/proto-parser';
import { recursiveModuleBundle, importNamespace } from '@osmonauts/ast';
import * as dotty from 'dotty';
import { TelescopeInput } from './index';
import { join, relative, dirname } from 'path';

// TODO move to store
export const getPackages = (store: ProtoStore) => {
    return store.getProtos().reduce((m, proto) => {
        if (proto.proto.package) {
            m[proto.proto.package] = m[proto.proto.package] || []
            m[proto.proto.package].push(proto);
        }
        return m;
    }, {});
}

export const getPackagesBundled = (store: ProtoStore) => {
    const objectified = {};
    const pkgs = getPackages(store);
    Object.keys(pkgs).forEach(key => {
        const files = pkgs[key];
        dotty.put(objectified, key, {
            pkg: key,
            files
        });
    });
    return objectified;

}

export const bundlePackages = (store: ProtoStore, input: TelescopeInput) => {
    const allPackages = getPackagesBundled(store);
    return Object.keys(allPackages).map(base => {
        const pkgs = allPackages[base];
        const bundleVariables = {};
        const bundleFile = join(input.protoDir, base, 'bundle.ts');
        const importPaths = [];
        parsePackage(pkgs, bundleFile, importPaths, bundleVariables);
        const body = recursiveModuleBundle(bundleVariables);
        return {
            bundleFile,
            importPaths,
            base,
            body
        };
    });
}

export const parsePackage = (obj, bundleFile, importPaths, bundleVariables) => {
    if (!obj) return;
    if (obj.pkg && obj.files) {
        obj.files.forEach(file => {
            createFileBundle(obj, file.filename, bundleFile, importPaths, bundleVariables);
            // TODO better namespacing for aminos/messages/encoding:
            // file.outFiles.forEach(outFile =>
            //     createFileBundle(obj, outFile.filename, bundleFile, importPaths, bundleVariables)
            // );
        });
        return;
    }
    Object.keys(obj).forEach(mini => {
        parsePackage(obj[mini], bundleFile, importPaths, bundleVariables);
    })
}


let counter = 0;
const createFileBundle = (obj, filename, bundleFile, importPaths, bundleVariables) => {
    let rel = relative(dirname(bundleFile), filename);
    if (!rel.startsWith('.')) rel = `./${rel}`;
    const variable = `_${counter++}`;
    importPaths.push(importNamespace(variable, rel));
    dotty.put(bundleVariables, obj.pkg + '.__export', true);
    dotty.put(bundleVariables, obj.pkg + '.' + variable, true);
}