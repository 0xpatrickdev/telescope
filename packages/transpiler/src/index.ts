import * as a from '@osmonauts/ast';
import * as t from '@babel/types';
import { extname, relative, dirname } from 'path';
import generate from '@babel/generator';
import { ProtoStore } from '@osmonauts/proto-parser';
import { parse, TelescopeParseContext } from './parse';
import {
    ProtoParseContext,
    AminoParseContext
} from '@osmonauts/ast';
export interface TelescopeInput {
    protoDir: string;
    outPath: string;
}

const UTILS = {
    AminoHeight: '@osmonauts/helpers',
    AminoMsg: '@cosmjs/amino',
    AminoTypes: '@cosmjs/stargate',
    decodeBech32Pubkey: '@cosmjs/amino',
    defaultRegistryTypes: '@cosmjs/stargate',
    encodeBech32PubKey: '@cosmjs/amino',
    fromBase64: '@cosmjs/encoding',
    fromBech32: '@cosmjs/encoding',
    fromDuration: '@osmonauts/helpers',
    fromHex: '@cosmjs/encoding',
    fromJsonTimestamp: '@osmonauts/helpers',
    fromTimestamp: '@osmonauts/helpers',
    GeneratedType: '@cosmjs/proto-signing',
    isSet: '@osmonauts/helpers',
    Long: '@osmonauts/helpers', // exports Long and also calls the magic Long code
    OfflineSigner: '@cosmjs/proto-signing',
    omitDefault: '@osmonauts/helpers',
    Registry: '@cosmjs/proto-signing',
    SigningStargateClient: '@cosmjs/stargate',
    toBase64: '@cosmjs/encoding',
    toDuration: '@osmonauts/helpers',
    toTimestamp: '@osmonauts/helpers',
};
export class TelescopeBuilder {
    store: ProtoStore;
    protoDir: string;
    outPath: string;

    constructor({ protoDir, outPath, store }: TelescopeInput & { store?: ProtoStore }) {
        this.protoDir = protoDir;
        this.outPath = outPath;

        this.store = store ?? new ProtoStore(protoDir);
        this.store.traverseAll();
    }


    buildProto(path) {
        const ref = this.store.findProto(path);

        const context = new TelescopeParseContext(
            ref, this.store
        );

        parse(context);

        // maybe we can do something like this to separate out the context..
        // parse(context1, {
        //     amino: true
        // });

        // parse(context2, {
        //     registry: true
        // });

        // messages = c.toObjectWithPartialMethods(this.mutations);
        // registry = c.createTypeRegistry(this.mutations);
        // json = c.toObjectWithJsonMethods(this.mutations);
        // toJSON = c.toObjectWithToJSONMethods(this.mutations);
        // fromJSON = c.toObjectWithFromJSONMethods(this.mutations);
        // encoded = c.toObjectWithEncodedMethods(this.mutations);

        // this.outFiles = [];


        const utils = Object.keys({
            ...context.amino.utils,
            ...context.proto.utils
        });

        const allImports = {};

        utils.forEach(util => {
            if (!UTILS.hasOwnProperty(util)) throw new Error('missing Util! ::' + util);
            allImports[UTILS[util]] = allImports[UTILS[util]] || [];
            if (!allImports[UTILS[util]].includes(util)) {
                allImports[UTILS[util]].push(util);
            }
        });

        Object.entries(context.amino.ref.traversed.parsedImports ?? {})
            .forEach(([filename, names]) => {
                const f = context.ref.filename;
                const rel = relative(dirname(f), filename);
                let importPath = rel.replace(extname(rel), '');
                if (!/\//.test(importPath)) importPath = `./${importPath}`;
                allImports[importPath] = allImports[importPath] || [];
                names.forEach(name => {
                    if (!allImports[importPath].includes(name)) {
                        allImports[importPath].push(name);
                    }
                })
            });

        const importStmts = Object.entries(allImports)
            .map(([pth, names]) => {
                return a.importStmt(names, pth);
            })

        const prog = []
            .concat(importStmts)
            .concat(context.body);

        const gen = generate(t.program(prog));
        return {
            gen,
            context
        }
    }
}

export default (input: TelescopeInput) => {
    const builder = new TelescopeBuilder(input);
    // const res = builder.buildProto('osmosis/gamm/v1beta1/tx.proto');
    // console.log(res);
};

