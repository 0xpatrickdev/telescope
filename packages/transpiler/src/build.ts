import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser';
import {
    importStmt,
    AminoParseContext,
    createAminoConverter,
    createCreateProtoType,
    createObjectWithMethods,
    createProtoEnum,
    createProtoType,
    makeAminoTypeInterface,
    ProtoParseContext,
    // client
    createClient,
    GenericParseContext,
    // registry 
    createTypeRegistry,
    createRegistryLoader,
    // helper
    createHelperObject,
} from '@osmonauts/ast';
import { extname, relative, dirname } from 'path';
import { ImportHash, ServiceMutation, ServiceQuery } from './types';
import { UTILS, getRelativePath, insertUniq, getRoot } from './utils';

export const buildAllImports = (context: TelescopeParseContext, allImports?: ImportHash) => {

    if (!allImports) allImports = {};

    const utils = Object.keys({
        ...context.amino.utils,
        ...context.proto.utils
    });

    utils.forEach(util => {
        if (!UTILS.hasOwnProperty(util)) throw new Error('missing Util! ::' + util);
        allImports[UTILS[util]] = allImports[UTILS[util]] || [];
        insertUniq(allImports[UTILS[util]], util);
    });

    const parsedImports: ImportHash = context.amino.ref.traversed.parsedImports ?? {};

    Object.entries(parsedImports)
        .forEach(([filename, names]) => {
            const importPath = getRelativePath(context.ref.filename, filename);
            allImports[importPath] = allImports[importPath] || [];
            names.forEach(name => insertUniq(allImports[importPath], name));
        });

    const importStmts = Object.entries(allImports)
        .map(([importPath, names]) => {
            return importStmt(names, importPath);
        })

    return importStmts;
}

export const getMutations = (mutations: ServiceMutation[]) => {
    return mutations.map((mutation: ServiceMutation) => {
        return {
            typeUrl: `/${mutation.package}.${mutation.message}`,
            TypeName: mutation.message,
            methodName: mutation.methodName
        }
    });
};

export const getAminoProtos = (mutations: ServiceMutation[], store: ProtoStore) => {
    return mutations.map(mutation => {
        const ref = store.findProto(mutation.messageImport);
        return store.get(ref, mutation.message).obj;
    });
};

export const getAminoImports = (mutations: ServiceMutation[]) => {
    return mutations.map(mutation => {
        return {
            import: mutation.messageImport,
            name: mutation.message
        };
    });
};

interface BuildBaseTSClass {
    includeEncode: boolean;
    includeDecode: boolean;
    includeFromPartial: boolean;
    includeFromJSON: boolean;
    includeToJSON: boolean;
}

export interface TelescopeParseContext {
    proto: ProtoParseContext;
    amino: AminoParseContext;
    store: ProtoStore;
    ref: ProtoRef;
    parsedImports: Record<string, any>;
    body: any[];
    mutations: ServiceMutation[];
    queries: any[];
    types: any[];
    options: any;
}

export const buildBaseTypeScriptClass = (
    context: TelescopeParseContext,
    name: string,
    obj: any,
    {
        includeEncode,
        includeDecode,
        includeFromPartial,
        includeFromJSON,
        includeToJSON
    }: BuildBaseTSClass = {
            includeEncode: true,
            includeDecode: true,
            includeFromPartial: true,
            includeFromJSON: true,
            includeToJSON: true
        }) => {

    context.body.push(createProtoType(name, obj));
    context.body.push(createCreateProtoType(name, obj));
    context.body.push(createObjectWithMethods(context.proto, name, obj));
};

export const getAminoRelativeDeps = (mutations: ServiceMutation[], filename: string) => {
    return getAminoImports(mutations)
        .map(imp => {
            const f = filename;
            const f2 = imp.import;
            if (f === f2) return;
            const rel = relative(dirname(f), f2);
            let importPath = rel.replace(extname(rel), '');
            if (!/\//.test(importPath)) importPath = `./${importPath}`;
            return {
                ...imp,
                importPath
            };
        })
        .filter(Boolean)
        .reduce((m, v) => {
            m[v.importPath] = m[v.importPath] ?? [];
            if (!m[v.importPath].includes(v.name)) {
                m[v.importPath].push(v.name);
            }
            return m;
        }, {});
};
export class TelescopeParseContext implements TelescopeParseContext {
    constructor(ref: ProtoRef, store: ProtoStore) {
        this.proto = new ProtoParseContext();
        this.amino = new AminoParseContext(
            ref, store
        );
        this.ref = ref;
        this.store = store;
        this.parsedImports = {};
        this.body = [];
        this.queries = [];
        this.mutations = [];
        this.types = [];
    }

    hasMutations() {
        return this.mutations.length > 0;
    }

    addType(name: string, obj: any) {
        this.types.push({
            name,
            obj
        });
    }
    addMutation(mutation: ServiceMutation) {
        this.mutations.push(mutation);
    }
    addQuery(query: ServiceQuery) {
        this.queries.push(query);
    }

    // build main Class with methods
    buildBase() {
        this.types.forEach(typeReg => {
            const { name, obj } = typeReg;
            if (obj.type === 'Type') {
                buildBaseTypeScriptClass(this, name, obj);
            } else if (obj.type === 'Enum') {
                this.body.push(createProtoEnum(name, obj));
            } else {
                throw new Error('buildBase(): unknown type');
            }
        })
    }

    buildRegistry() {
        this.body.push(createTypeRegistry(getMutations(this.mutations)));
    }
    buildRegistryLoader() {
        this.body.push(createRegistryLoader());
    }
    buildAminoInterfaces() {
        const protos = getAminoProtos(this.mutations, this.store);
        protos.forEach(proto => {
            this.body.push(makeAminoTypeInterface({
                context: this.amino,
                proto
            }));
        });
    }
    buildAminoConverter() {
        this.body.push(createAminoConverter({
            name: 'AminoConverter',
            context: this.amino,
            root: this.ref.traversed,
            protos: getAminoProtos(this.mutations, this.store)
        }));
    }
    buildHelperObject() {
        // add methods
        this.body.push(createHelperObject({
            name: 'MessageComposer',
            mutations: getMutations(this.mutations)
        }));
    }
}