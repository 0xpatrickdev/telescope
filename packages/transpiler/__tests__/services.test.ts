import {
    ProtoParseContext,
    AminoParseContext
} from '@osmonauts/ast';
import generate from '@babel/generator';
import { ProtoRef, ProtoStore, traverse, getNestedProtoGeneric, parseProto } from '@osmonauts/proto-parser'
import { camel, snake } from 'case';
import { parse, TelescopeParseContext } from '../src/parse';

const store = new ProtoStore('');
store.protos = [];
const addRef = ({ filename, content }) => {
    const ref = {
        absolute: filename,
        filename,
        proto: parseProto(content)
    };
    store.protos.push(ref);
};
addRef({
    filename: 'cosmology/example/a.proto',
    content: `
syntax = "proto3";

package cosmology.finance;
option go_package = "github.com/cosmology-finance/go";

message MsgTypePackageA {
    string address = 1;
}

`});
addRef({
    filename: 'cosmology/example/b.proto',
    content: `
syntax = "proto3";

package cosmology.finance;
option go_package = "github.com/cosmology-finance/go";

import "cosmology/example/a.proto";

message MsgTypePackageB {
    string name = 2;
}

message MsgTypePackageBResponse {}

service Msg {
    rpc ActionFromPackageBInner(MsgTypePackageB) returns (MsgTypePackageBResponse);
}

`});
addRef({
    filename: 'cosmology/example/c.proto',
    content: `
syntax = "proto3";
package cosmology.finance;
option go_package = "github.com/cosmology-finance/go";

import "cosmology/example/b.proto";

message MsgTypePackageC {
    string                              address        = 1;
    cosmology.finance.MsgTypePackageB     awesome    = 2;
}

service Msg {
    rpc ActionFromPackageC(MsgTypePackageC) returns (MsgTypePackageCResponse);
    rpc ActionFromPackageB(MsgTypePackageB) returns (MsgTypePackageBResponse);
}
message MsgTypePackageCResponse {}

service Query {
    // Config returns the current app config.
    rpc Config(QueryConfigRequest) returns (QueryConfigResponse) {}
}

// QueryConfigRequest is the Query/Config request type.
message QueryConfigRequest {}

// QueryConfigRequest is the Query/Config response type.
message QueryConfigResponse {

// config is the current app config.
Config config = 1;
}
  

`});

store.traverseAll();

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}
const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

describe('cosmology/example/c', () => {
    store.traverseAll();

    const ref = store.findProto('cosmology/example/c.proto');
    const context: TelescopeParseContext = {
        proto: new ProtoParseContext(),
        amino: new AminoParseContext(
            ref, store
        ),
        ref,
        store: store,
        parsedImports: {},
        body: [],
        mutations: [],
        queries: [],
        types: []
    };

    // aggregate service information
    // - [ ] get dependent objects of service and their paths
    // aggregate amino information
    // - [ ] get all aminos into one bundle per package!

    it('works', () => {
        parse(context);
        expect(context.mutations).toMatchSnapshot();
        expect(context.queries).toMatchSnapshot();
    });

});
