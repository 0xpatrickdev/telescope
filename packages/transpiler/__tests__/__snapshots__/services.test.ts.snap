// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cosmology/example/c works 1`] = `
Array [
  Object {
    "comment": null,
    "message": "MsgTypePackageC",
    "messageImport": "cosmology/example/c.proto",
    "methodName": "ActionFromPackageC",
    "package": "cosmology.finance",
    "response": "MsgTypePackageCResponse",
    "responseImport": "cosmology/example/c.proto",
  },
  Object {
    "comment": null,
    "message": "MsgTypePackageB",
    "messageImport": "cosmology/example/b.proto",
    "methodName": "ActionFromPackageB",
    "package": "cosmology.finance",
    "response": "MsgTypePackageBResponse",
    "responseImport": "cosmology/example/b.proto",
  },
]
`;

exports[`cosmology/example/c works 2`] = `
Array [
  Object {
    "comment": "Config returns the current app config.",
    "message": "QueryConfigRequest",
    "messageImport": "cosmology/example/c.proto",
    "methodName": "Config",
    "package": "cosmology.finance",
    "response": "QueryConfigResponse",
    "responseImport": "cosmology/example/c.proto",
  },
]
`;
